### PR TITLE
chore(config): simplify and prep providers a bit

### PR DESCRIPTION
**Pull request by lukesteensen**
This makes a few small changes to providers to prep for remote config work:
1. Moves them within the config module
2. Removes the unused `ProviderDescription` concept
3. Moves more towards an always-used strategy-style implementation rather than a separate path

Overall this makes it easier to slot in an enterprise strategy, where we pull a provider from the enterprise config itself (i.e. no need for a separate `[provider]` block). I believe the only behavior change here is that the debug log of which provider is used is no longer conditional.

### CODE REVIEW

File: src/config/loading/mod.rs [hunk 1]
```diff
@@ -151,14 +149,17 @@ pub async fn load_from_paths_with_provider_and_secrets(
...
-    if let Some(mut provider) = builder.provider {
-        builder = provider.build(signal_handler).await?;
-        debug!(message = "Provider configured.", provider = ?provider.provider_type());
-    }
+    // This is where we conceptually go from the config as it is on disk to whatever the "full"
+    // config is going to be. In the simplest case these are the same thing, and all
+    // sources/transforms/sinks are defined directly. In the case of a `[provider]` block or other
+    // remote configuration, we expect to get the actual topology from elsewhere. It would be a bit
+    // clearer if on-disk config and full topology config were better differentiated in the type
+    // system, but for now we essentially just replace one `ConfigBuilder` with another.
```
Comment (spencergilbert) [inline src/config/loading/mod.rs:159]: Do we have planned/future-work to differentiate these via types?

Comment (lukesteensen) [reply src/config/loading/mod.rs:159]: This is something I initially looked at doing as part of this work but couldn't justify the extra time. I have some notes I'll write up into an issue that we can prioritize separately.

**Review by spencergilbert — APPROVED** [review]
Changes look straightforward, but I haven't interacted with the providers much - definitely worth getting a 👍 from @leebenson

**Comment by jszwedko** [comment]
@lukesteensen do we still want this or can we close it?

**Comment by lukesteensen** [comment]
@jszwedko it's not super important, but maybe some value in the refactoring still? Maybe @leebenson has feelings one way or the other.

**Comment by jszwedko** [comment]
@lukesteensen is this still useful outside of remote config (given that is moving)?

**Comment by jszwedko** [comment]
I'll close this assuming it isn't relevant, but we can reopen if it is.

**Comment by lukesteensen** [comment]
Without the remote config follow up, this is really just refactoring to make things a bit clearer. It's still a positive change in my opinion, but not necessary.

**Comment by jszwedko** [comment]
> Without the remote config follow up, this is really just refactoring to make things a bit clearer. It's still a positive change in my opinion, but not necessary.

👍 I'll leave it here as a reference at least since it doesn't seem likely we'd pick it back up without some more focus on Vector's providers. Unless you feel like pushing it across the line now 😄

**Pull request closed without merging**